msal_extensions/token_cache.py

```diff
@@ -1,156 +1,119 @@
 import os
-import sys
 import warnings
 import time
 import errno
-import msal
-from .cache_lock import CrossPlatLock
+import logging
 
-if sys.platform.startswith('win'):
-    from .windows import WindowsDataProtectionAgent
-elif sys.platform.startswith('darwin'):
-    from .osx import Keychain
+import msal
 
-def _mkdir_p(path):
-    """Creates a directory, and any necessary parents.
+from .cache_lock import CrossPlatLock
+from .persistence import (
+    _mkdir_p, FilePersistence,
+    DataProtectionPersistence, KeychainPersistence, LibsecretPersistence)
 
-    This implementation based on a Stack Overflow question that can be found here:
-    https://stackoverflow.com/questions/600268/mkdir-p-functionality-in-python
 
-    If the path provided is an existing file, this function raises an exception.
-    :param path: The directory name that should be created.
-    """
-    try:
-        os.makedirs(path)
-    except OSError as exp:
-        if exp.errno == errno.EEXIST and os.path.isdir(path):
-            pass
-        else:
-            raise
+logger = logging.getLogger(__name__)
 
+class PersistedTokenCache(msal.SerializableTokenCache):
+    """A token cache using given persistence layer, coordinated by a file lock."""
 
-class FileTokenCache(msal.SerializableTokenCache):
-    """Implements basic unprotected SerializableTokenCache to a plain-text file."""
-    def __init__(self,
-                 cache_location,
-                 lock_location=None):
-        super(FileTokenCache, self).__init__()
-        self._cache_location = cache_location
-        self._lock_location = lock_location or self._cache_location + '.lockfile'
+    def __init__(self, persistence, lock_location=None):
+        super(PersistedTokenCache, self).__init__()
+        self._lock_location = (
+            os.path.expanduser(lock_location) if lock_location
+            else persistence.location + ".lockfile")
+        _mkdir_p(os.path.dirname(self._lock_location))
+        self._persistence = persistence
         self._last_sync = 0  # _last_sync is a Unixtime
+        self.is_encrypted = False
 
-        self._cache_location = os.path.expanduser(self._cache_location)
-        self._lock_location = os.path.expanduser(self._lock_location)
-
-        _mkdir_p(os.path.dirname(self._lock_location))
-        _mkdir_p(os.path.dirname(self._cache_location))
-
-    def _needs_refresh(self):
-        # type: () -> Bool
-        """
-        Inspects the file holding the encrypted TokenCache to see if a read is necessary.
-        :return: True if there are changes not reflected in memory, False otherwise.
-        """
+    def _reload_if_necessary(self):
+        # type: () -> None
+        """Reload cache from persistence layer, if necessary"""
         try:
-            updated = os.path.getmtime(self._cache_location)
-            return self._last_sync < updated
+            if self._last_sync < self._persistence.getmtime():
```

Comment:
I think this needs more testing. I believe the .NET implementers found that timestamps are not granular enough to ensure locking.

Reply:
Here the timestamp is only used to determine whether the data copy in memory becomes stale (and it seems we won't have other ways to detect that).

This timestamp check is NOT used for locking mechanism. The lock has already been acquired when the app reaches this line of code.